Src/IronPython.Modules/resource.cs

```diff
@@ -0,0 +1,427 @@
+        public object? ru_majflt => this[7];
+        public object? ru_nswap => this[8];
+        public object? ru_inblock => this[9];
+        public object? ru_oublock => this[10];
+        public object? ru_msgsnd => this[11];
+        public object? ru_msgrcv => this[12];
+        public object? ru_nsignals => this[13];
+        public object? ru_nvcsw => this[14];
+        public object? ru_nivcsw => this[15];
+
+        public override string __repr__(CodeContext/*!*/ context) {
+            return string.Format("resource.struct_rusage("
+                + "ru_utime={0}, "
+                + "ru_stime={1}, "
+                + "ru_maxrss={2}, "
+                + "ru_ixrss={3}, "
+                + "ru_idrss={4}, "
+                + "ru_isrss={5}, "
+                + "ru_minflt={6}, "
+                + "ru_majflt={7}, "
+                + "ru_nswap={8}, "
+                + "ru_inblock={9}, "
+                + "ru_oublock={10}, "
+                + "ru_msgsnd={11}, "
+                + "ru_msgrcv={12}, "
+                + "ru_nsignals={13}, "
+                + "ru_nvcsw={14}, "
+                + "ru_nivcsw={15})",
+                this.Select(v => PythonOps.Repr(context, v)).ToArray());
+        }
+
+    }
+
+    private static long? GetLimitValue(System.Collections.IEnumerator cursor) {
+        if (!cursor.MoveNext() || !PythonOps.TryToIndex(cursor.Current, out BigInteger lim))
+            return null;
+
+        long rlim = checked((long)lim);
+        if (rlim < 0 && RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            rlim -= long.MinValue;
+        return rlim;
+    }
+
+    private static object LimitsArgError()
+        => LightExceptions.Throw(PythonOps.ValueError("expected a tuple of 2 integers"));
+
+    private static object GetPInvokeError() {
+        int errno = Marshal.GetLastWin32Error(); // despite its name, on Posix it retrieves errno set by the last p/Invoke call
+        return LightExceptions.Throw(PythonOps.OSError(errno, PythonNT.strerror(errno)));
+    }
+
+    private static object ToPythonInt(this ulong value)
+        => unchecked((long)value).ToPythonInt();
+
+    private static object ToPythonInt(this long value)
+        => value is <= int.MaxValue and >= int.MinValue ? (int)value : (BigInteger)value;
+
+    private struct rlimit {
+        // /usr/include/x86_64-linux-gnu/bits/resource.h
+        // /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/resource.h
+
+        public ulong rlim_cur;
+        public ulong rlim_max;
+    }
+
+#pragma warning disable CS0649 // Field is never assigned to, and will always have its default value
+    internal struct rusage {
+        // /usr/include/x86_64-linux-gnu/bits/types/struct_rusage.h
+        // /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/resource.h
+
+        public Mono.Unix.Native.Timeval ru_utime;  // user CPU time used
+        public Mono.Unix.Native.Timeval ru_stime;  // system CPU time used
```

Comment:
I am not happy with using a type from _Mono.Unix_, here; well, I am not happy with _Mono.Unix_ in general: it does not work correctly on macOS. Although there are no issues with `Mono.Unix.Native.Timeval`, I think that by avoiding using _Mono.Unix_ altogether, the dependency could be removed at some point (at least on macOS, it seems to work OK on Linux).

Reply:
Since _Mono.Unix_ is proven not to work correctly in all cases, assuming that any call through this library will work as expected without any testing is asking for trouble. I myself got burned on this assumption a few times. So in practical terms, _Mono.Unix_ must be considered untested (on macOS), and we need good test coverage regardless whether the call goes through the library or directly through P/Invoke. The advantage of P/Invoke is that it is readable/debuggable; I can check every signature and types used for correctness. On top of that, quite some of the missing functionality is not supported by _Mono.Unix_ anyway (example: `resource`), and what is supported is usually also (and better) accessible through the regular .NET API. Further, in some cases that Mono does offer some additional and needed functionality and it happens to work, it tweaks the interface slightly to make it a little bit more convenient to use from C#. But this is not the best for IronPython: CPython's modules expose the C lib API at its raw low level, and accessing C lib directly though P/Invoke guarantees that the implementation does what CPython would do, no need to reverse-engineer Mono API. All in all, I think there will be only a small number of cases where _Mono.Unix_ will be practical. This "small number" is not a scientific number, more my gut feeling, we'll see how it is after implementing all the missing modules. But everything is pointing in the direction that there will be "a bunch  of platform specific P/Invokes" which will have to be covered by tests.

Reply:
> Although there are no issues with `Mono.Unix.Native.Timeval`

I have spoken too soon. On Darwin, `timeval` is `struct { long, int }`. `Mono.Unix.Native.Timeval` is `struct { long, long }` (Linux layout).